### ISSUE TITLE
Working on high res .tif satellite images

### ISSUE THREAD

**Issue by ash1995** [matterport/Mask_RCNN#1197]
I've been trying to utilise the Mask RCNN to read satellite geotiff images from the SpaceNet dataset for building detection. These high res .tif  images have 4 bands namely RGB and Near infrared (NIR). I've tried cutting out the NIR channel and converting it to jpeg files before training the Mask RCNN but the performance was quite bad. Now this is most probably due to information loss from the conversion from tif to jpeg as I have observed the jpeg images are not nearly as clear as the tif images. 

So my question is, has anyone used the Mask RCNN for high res tif images and if so how did you go about doing it?

**Comment by yghlc** [comment]
I did last month. How bad are you results? total wrong or accuracies are unexpected? I have tried Mask_RCNN on Planet images (with resolution = 3m and four bands).  I tried images with four bands and three bands. I got reasonable results but the accuracies are not as high as I expected. Maybe I need more time to go though the codes.

**Comment by ash1995** [comment]
I fixed the image darkness problem by cutting out the NIR band and using adaptive histogram equalization to improve the contrast in the image but the model still performs poorly.

So for context the below images shows the ground truth bounding boxes and segmentation polygons.

![screen shot 2018-12-30 at 10 40 19 pm](https://user-images.githubusercontent.com/32295036/50550157-02944a80-0c84-11e9-8b60-cfeb1bb8e0a0.jpg)


<img width="606" alt="screen shot 2018-12-30 at 10 37 27 pm" src="https://user-images.githubusercontent.com/32295036/50550160-0aec8580-0c84-11e9-832d-c92023038b83.png">


 After training the Mask RCNN on my data, I'm getting detections like this:


![screen shot 2018-12-30 at 10 45 52 pm](https://user-images.githubusercontent.com/32295036/50550200-b1d12180-0c84-11e9-9939-9b13465e873a.jpg)


Most of the time, the bounding boxes appear in random locations and somewhat random sizes. The performance isn't consistent either. Every inference yields random results. So yeah I would say totally wrong but I'm not able to pinpoint what's causing this.

**Comment by yghlc** [comment]
Sould be some bugs in your codes.

**Comment by ash1995** [comment]
The issue is I'm not sure where I should start debugging. I haven't changed the default implementation much and I had converted my annotations so that it matches the standard COCO format for which matterport has a template.. In your opinion, what are the plausible reasons for this happening? Do you think it's more of a software/coding bug or is it more related to optimizing hyper parameters?

**Comment by yghlc** [comment]
In image processing, remote sensing images are not different from other images such as COCO images Expect their multi-bands and large size. 
You may try COCO dataset on your implementation first, if it works, then you should check the codes of converting data format. Otherwise, you should check what you change to the default implementation. 

I did not use the COCO format. Instead, I wrote codes for pre-processing and post-processing and implemented a new "road_mask" for mas_rccn.

**Comment by bilgan** [comment]
How did you find those white polygons in the second image? Is it your label?

**Comment by bilgan** [comment]
How did you find those white polygon in second image? Is it your label?

**Comment by ash1995** [comment]
@bilgan The polygon pixel coordinates were provided with the dataset I'm using so it was just a matter of drawing it on the image.

**Comment by ash1995** [comment]
So it turns out while loading the model for inference I was randomly initialising certain layers that were originally excluded via the exclude parameter. Once that was rectified, my model began working.

**Comment by MrDemeanor** [comment]
@ash1995 Do you have this project in a public repo? I am doing the same task but I'm having trouble getting my model to train correctly. Also, what was your config? (epochs, steps per epoch, max/min size, etc?)

**Comment by ashnair1** [comment]
@MrDemeanor You can checkout my repo [here](https://github.com/ashnair1/Mask-RCNN-for-Off-Nadir-Building-Detection). I used Mask RCNN to train on the SpaceNet Off Nadir dataset.

**Comment by MrDemeanor** [comment]
@ash1995 Thank you very much! Checking it out now.

**Comment by fabioaraujopt** [comment]
start by doing something as balloon recognition and then go forward two train for a single class. than go to more classes!
https://engineering.matterport.com/splash-of-color-instance-segmentation-with-mask-r-cnn-and-tensorflow-7c761e238b46

**Comment by Amrimn** [comment]
@ashnair1 that for the repository I'm also using satellite images. for the prediction i can't give a big image ( 8000*7000) i get memory error when the detection start( **model.detect([image], verbose=0)[0]**). With small image (4000*4000) everything is fine.
I am trying to crops the image an reconstruct the result of the prediction.
Could you please give me more insight on the above issues? I thank you a lot in advance for your collaboration.

**Comment by ashnair1** [comment]
@Amrimn The data I had used was the Spacenet Off Nadir dataset. This dataset consists of image tiles of size 900 x 900 pixels. I have unfortunately not tried out inference on really large images yet using that repo. As you said, you could crop the image, get and save the inference results from the model and then later merge the results. Definitely doable.

**Comment by Amrimn** [comment]
@ashnair1 In fact, I crop the image and  for each crop, I get the dictionary (result) of the below arrays:
The dict contains:
-masks: [H, W, N] instance binary masks.
-rois: [N, (y1, x1, y2, x2)] detection bounding boxes
-class_ids: [N] int class IDs
-scores: [N] float probability scores for the class IDs
 
I manage to reconstruct the mask as I know the position of the crop in the big image(8000*7000). However, for the rest of the dictionary elements I can't figure out how to group them properly in one array each with respect to the original large image.

**Comment by ashnair1** [comment]
@Amrimn  If your goal is just to get the overall results, you could try
1. Run inference on a crop (like 0_0.jpg)
2. Save the inference result as a jpg (like 0_0_inf.jpg)
3. Merge all the inference jpgs together to get back the big image

**Comment by hamzahkhan** [comment]
@ashnair1 How do you interpret the mAP value reported i.e. 30.4? Is it good bad?

**Comment by aldinorizaldy** [comment]
> start by doing something as balloon recognition and then go forward two train for a single class. than go to more classes!
> https://engineering.matterport.com/splash-of-color-instance-segmentation-with-mask-r-cnn-and-tensorflow-7c761e238b46

I did the similar steps with balloon for drone RGB imageries. The results are good enough. Now I'm gonna use satellite imageries with more than 3 bands. How to deal with that? Do I need to change something in the architecture since balloon (and the original mask rcnn) only use 3 channels RGB. Also can I use pre-trained COCO as well?

**Comment by fabioaraujopt** [comment]
About more than 3 channels, read this:

https://github.com/matterport/Mask_RCNN/wiki#training-with-rgb-d-or-grayscale-images

aldinorizaldy <notifications@github.com> escreveu no dia terça, 7/07/2020
à(s) 02:34:
